### PR TITLE
fix(payment): atomic CAS for webhook watermark & linked-token migration

### PR DESCRIPTION
**Pull request by billmal071**
Follow-up to #411 addressing CodeRabbit findings that landed after the subscription-webhook code was merged into `develop-v1.2.0`.

## 1. Watermark update is now an atomic compare-and-swap (Major)
`src/payment/subscription-webhook.service.ts` — `applyAction()`

The out-of-order guard previously read `subscription.lastEventAt`, compared it to the event timestamp in JS, then did a separate `update` by `id` — a read-then-write TOCTOU race where two concurrent deliveries could both read the old watermark, both pass, and both write (older clobbering newer). The comparison now lives INSIDE the write via a conditional `updateMany`, and we branch on the affected-row count:

- `count === 1` → the guard matched; the state change and the advanced `lastEventAt` are committed in the same atomic statement.
- `count === 0` → a newer watermark is already stored (stale / lost the race); nothing is written and the WebhookEvent is recorded as `skipped` (still 200).

Events with no derivable timestamp still apply (unconditional `update`). Idempotency (WebhookEvent unique on `externalEventId`) and the linked-token flow are unchanged.

## 2. Equal-timestamp distinct events are no longer dropped (Major)
`prisma/schema.prisma` comment + the watermark comparison

The old `>=` comparison treated a DISTINCT event arriving at the same millisecond as a stale duplicate. The CAS WHERE now uses `lastEventAt IS NULL OR lastEventAt <= eventAt`, so same-instant distinct notifications APPLY in arrival order and only strictly-older events skip. True duplicates remain handled by the `WebhookEvent (platform, externalEventId)` idempotency layer. The chosen semantics are documented on `Subscription.lastEventAt` in the schema and in the `applyAction` doc comment.

## 3. Rejection tests assert HTTP status 400 (Minor)
`src/payment/apple-notification.spec.ts`

The bundleId and array-header rejection cases now assert the thrown `HttpException` carries status 400 (via `getStatus()`), so a regression to a 500 is caught.

## Bonus: linked-token migration is compare-and-swap
The same TOCTOU class in the linked-token migration (validate old token in memory, then update by `id`) is fixed in both `subscription-webhook.service.ts` (`resolveGoogleSubscription`) and `payment.service.ts` (`migrateGoogleLinkedToken`): the update is guarded on the still-stored old token (`updateMany`), and `count === 0` re-resolves / no-ops instead of mis-mapping the row.

## Tests
Added/updated: concurrent old/new CAS delivery (older loses → skipped, newer applies), equal-timestamp distinct events both apply, no-prior-watermark apply, and linked-token migration CAS lost-race. `pnpm build`, `pnpm lint` clean; `pnpm test -- src/payment` → 98 passing.

## Out of scope
Deferred per PR #411 discussion (CodeRabbit withdrew / logged as follow-ups): Pub/Sub OIDC bearer-token validation, `fetchLinkedPurchaseToken` error propagation, and requiring `result.success` before activation (RTDN notification type is the authoritative signal).

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved subscription webhook handling to safely ignore out-of-order events using atomic watermark gating (including consistent handling for equal timestamps).
  * Prevented concurrent Google purchase-token migrations from clobbering a newer winning token.
  * Strengthened failure behavior for purchase verification by properly propagating conflict errors.
  * Improved Apple notification error responses with clearer HTTP 400 messaging.
* **Tests**
  * Expanded assertions for stale delivery, watermark edge cases, concurrent races, and token-migration retry/idempotency behavior.
  * Tightened malformed and bundle mismatch notification validation.

### CODE REVIEW

File: src/payment/payment.service.ts [hunk 1]
```diff
@@ -230,10 +230,21 @@ export class PaymentService {
...
+    const res = await this.prisma.subscription.updateMany({
+      where: { id: existing.id, purchaseToken: linkedToken },
       data: { purchaseToken: newToken },
     });
+    if (res.count === 0) {
+      this.logger.log(
+        `Skipped Google purchaseToken migration for user ${userId.substring(0, 8)} (already migrated concurrently)`,
+      );
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (billmal071) [reply src/payment/payment.service.ts:273]: Fixed in 8504212.

The zero-row linked-token CAS is no longer treated as unconditionally benign. On `count === 0` we now re-read the row by `id` and:
- continue only when it already holds `newToken` (our own write / an idempotent retry), or
- throw `ConflictException` when a concurrent delivery installed a **different** token — so the current delivery aborts before the id-only upsert and the concurrent winner is preserved (never clobbered).

Additionally, the final subscription upsert (`upsertSubscriptionWithExpiry`) now performs a token-guarded `updateMany` (`WHERE id AND purchaseToken = <observed token>`) instead of an unconditional `update`-by-id; if the guard misses it re-reads and keeps the concurrent winner rather than overwriting it. Added tests: a CAS-conflict case asserting the other token survives, and an idempotent same-token repeat that still succeeds. `pnpm test -- src/payment` green (101 passing).

File: src/payment/subscription-webhook.service.ts [hunk 2]
```diff
@@ -722,11 +736,27 @@ export class SubscriptionWebhookService {
...
+   * Ordering semantics (see also `prisma/schema.prisma` `Subscription.lastEventAt`):
+   * apply when there is no stored watermark yet OR the stored watermark is <= this
+   * event (`lte`). Equality APPLIES rather than skips: two DISTINCT notifications
+   * can share a millisecond, and true duplicates are already filtered upstream by
+   * the `WebhookEvent (platform, externalEventId)` idempotency layer, so a
+   * same-instant event reaching here is always a distinct one that must be
+   * processed in arrival order. Only a STRICTLY greater stored watermark skips.
+   *
+   * When `eventAt` is null the event has no derivable timestamp; ordering cannot
+   * be established, so it is applied unconditionally (never blocked).
```
[the remark this answers is not in the document]
Comment (billmal071) [reply src/payment/subscription-webhook.service.ts:768]: Addressed in 8504212 — corrected the doc, kept `<=` (option 2).

You are right that two conflicting same-millisecond events both pass the `lte` guard and resolve last-write-wins in DB-commit order, which need not equal arrival order. Rather than over-claim arrival ordering, I removed the "must be processed in arrival order" wording and the doc now states the honest limitation explicitly: same-timestamp conflicts are applied in DB-commit order (last-write-wins), not serialized by arrival.

I deliberately did **not** add a monotonic per-subscription arrival key or per-subscription write serialization: that requires a schema migration / locking infra that is not warranted here, because Apple and Google emit one notification per state transition, each with a distinct timestamp — conflicting same-millisecond lifecycle events for a single subscription do not occur in practice (and true duplicates are already filtered upstream by the `WebhookEvent (platform, externalEventId)` idempotency layer). Equality still APPLIES (not skips) so distinct same-instant events are never dropped.

Test added: two conflicting same-timestamp Apple events (DID_RENEW then EXPIRED) both apply (`processed`), documenting the last-write-wins behavior. `pnpm test -- src/payment` green (101 passing).
